newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/ShapePreview.js

```diff
@@ -0,0 +1,213 @@
+// @flow
+import * as React from 'react';
+
+type Vertex = {
+  x: number,
+  y: number,
+};
+
+type Props = {|
+  shape: string,
+  dimensionA: number,
+  dimensionB: number,
+  offsetX: number,
+  offsetY: number,
+  polygonOrigin: string,
+  vertices: Array<Vertex>,
+  width: number,
+  height: number,
+  onMoveVertex: (index: number, newX: number, newY: number) => void,
+|};
+
+type State = {|
+  draggedVertex: ?Vertex,
+  draggedIndex: number,
+|};
+
+export default class ShapePreview extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = { draggedVertex: null, draggedIndex: -1 };
+  }
+
+  _svg: any;
+
+  _onVertexDown = (vertex: Vertex, index: number) => {
+    if (this.state.draggedVertex) return;
+    this.setState({ draggedVertex: vertex, draggedIndex: index });
+  };
+
+  _onMouseUp = () => {
+    const draggingWasDone = !!this.state.draggedVertex;
+    const { draggedVertex, draggedIndex } = this.state;
+    this.setState(
+      {
+        draggedVertex: null,
+      },
+      () => {
+        if (draggingWasDone)
+          this.props.onMoveVertex(
+            draggedIndex,
+            Math.round(draggedVertex ? draggedVertex.x : 0),
+            Math.round(draggedVertex ? draggedVertex.y : 0)
+          );
+      }
+    );
+  };
+
+  _onMouseMove = (event: any) => {
+    const { offsetX, offsetY, polygonOrigin, width, height } = this.props;
+    const { draggedVertex } = this.state;
+    if (!draggedVertex) return;
+
+    const pointOnScreen = this._svg.createSVGPoint();
```

Comment:
I recognize this code ;) I think at some point we'll have to factor this

Reply:
Yeah, I see what you did there, coordinate systems between different elements is a pain :/